src/mminou.c

```diff
@@ -587,6 +588,26 @@ void printLongLine(const char *line, const char *startNextLine, const char *brea
     }
   } // if (breakMatch1[0] == '\"')
 
+  // TeX mode
+  // The TeX mode is intended not to break inside curly brace scopes.
+  // Whenever we are inside a scope of curly braces, we change a space to
+  // ASCII 3 to prevent matching it.  The reverse is done in the print2()
+  // function, where all ASCII 3's are converted back to space.
+  if (breakMatch1[0] == ' ') {
+    i = 0;
+    // k counts the scope level we are in.
+    k = 0;
+    while (multiLine[i]) {
+      // We enter a non "\{" scope.
+      if (multiLine[i] == '{' && multiLine[i - 1] != '\\') k++;
+      // We escape a non "\}" scope.
+      if (multiLine[i] == '}' && multiLine[i - 1] != '\\') k--;
+      // If k > 0 then we are inside a scope.
+      if (multiLine[i] == ' ' && k > 0) multiLine[i] = QUOTED_SPACE;
+      i++;
+    }
+  }
+
```

Comment:
The function is `printLongLine`, I think it's used to print out pretty much everything.
I don't see any check that we actually are in TeX mode. Is this code inoffensive in e.g. ASCII mode?

Reply:
You're right. 

So, to address this I'm thinking about imitating the HTML again. The curly braces issue affects only these two printed lines:

https://github.com/metamath/metamath-exe/blob/ae35c1b2931be8f9683f8bf66d1d10c9b56db706/src/mmwtex.c#L3046-L3099

So the idea would be to change `breakMatch` (the third argument of `printLongLine`) from a space to a special character, to be sure that we are in the above cases only. Then we change the special charater to a space again so that the following rewrapping algorithm works as expected:
https://github.com/metamath/metamath-exe/blob/ae35c1b2931be8f9683f8bf66d1d10c9b56db706/src/mminou.c#L567-L568

What do you think?

Reply:
What about only performing that treatment within this condition?
```
if (!g_htmlFlag) { // LaTeX 
   ...
}
```
it's supposed to be a global flag switching the output mode between HTML and LaTeX.
What I'm afraid of, is that it might still be set when we print to the standard output.

Reply:
You mean converting all spaces into `QUOTED_SPACE` inside the `if (!g_htmlFlag) { // LaTeX` condition instead of the mminou.c file? That might be not a bad idea.

Reply:
To be honest I'm not quite sure, I certainly did not take time to fully analyze the code...

Reply:
I attempted to put the brace count inside the `if (!g_htmlFlag) { // LaTeX` condition and it seems to work. But before eventually committing it I have a question:

With this new apporach the `line` argument of `printLongLine` contains some `QUOTED_SPACE` characters and that triggers bug check 1514. The reason why it gets triggered it pretty straightforward since `bug(1514)` checks exacly that:

https://github.com/metamath/metamath-exe/blob/ae35c1b2931be8f9683f8bf66d1d10c9b56db706/src/mminou.c#L562-L566

I don't understand the point of this bug check, it seems that its purpose is to make sure that no `QUOTED_SPACE` is printed, but, in my eyes it fails at checking what it claims to check. According to the `// HTML mode` comment, all occurrences of  `QUOTED_SPACE` get converted back into spaces inside the `print2()` function, so wouldn't be more useful to put this check in `print2()` instead? Maybe at the end of all operations, right before printing?

So to make it work I had to remove the above lines and the resulting TeX looks fine.

Reply:
What I understand is the following:
The algorithm wants to prevent line breaking at certain places (spaces within double-quote `"` quotation). To do so, it first replaces all such spaces by `QUOTED_SPACE`, then performs the line breaking, which acts on the remaining spaces, and finally switches back the `QUOTED_SPACE ` to normal spaces. 

This is supposed to be transparent, in the sense that space in the input results in spaces or line breaks in the output, and no spaces are added.

If there are `QUOTED_SPACE` in the original string, even before the whole process, those would be turned into normal spaces at the end of the processing, so spaces are "added" and the above contract is not respected. This makes `QUOTED_SPACE` a special character which should not be present in the input (and therefore the bug 1514).

Here, your (our) changes would expose some inner working of `printLongLine` to the caller of the function, and its contract becomes:
- the function breaks long lines into suitable mutli-lines according to the rules, etc., as before,
- any `QUOTED_SPACE` character is replaced by a space in the output, and cannot be a line break position.

We're using this "new" functionality in the caller by putting `QUOTED_SPACE` where LaTeX does not want line breaks to appear.
And because the `QUOTED_SPACE` thing is now part of the function's pre/post conditions, it is not a reserved character anymore and we can remove the bug check safely.

If this is agreed and we move on with the change, it would be nice to put some of these thoughts in or around the function's comment!

Reply:
Yep, everything you said aligns with my thoughts, so it seems we are on the same page. Unless somebody else have objections/other proposals, I'm going to commit the new proposed solution soon (it basically subsumes the old one).

Reply:
I committed the new proposal. To keep topics in order, I mark this thread as resolved since we reached an agreement. Open a new one for further discussion.